### PR TITLE
[GCP] Add document for minimal permission

### LINKED ISSUE #2086: [Question] Strong Minimal Cloud Permissions with GCP
Hi, according to [the documentation](https://skypilot.readthedocs.io/en/latest/cloud-setup/cloud-permissions.html#gcp) the following roles must be granted to each account using SkyPilot:
```
roles/browser
roles/compute.admin
roles/iam.serviceAccountAdmin
roles/iam.serviceAccountUser
roles/serviceusage.serviceUsageConsumer
roles/storage.admin
```

Some of these roles are quiet strong (most notably `iam.serviceAccountAdmin`), I'm curious about the rationale behind these requirements. My issue with it is the following: our GCP project is shared between multiple developers, each developer is granted a fine-grained permission set depending on what is needed. We want to assess if it's possible for all developers to use SkyPilot without granting excessive additional permissions.

Initially I thought that `roles/iam.serviceAccountAdmin` was only needed once, to create the SkyPilot service account, and that users can all start and use VMs using this existing service account. However it appears that `roles/iam.serviceAccountAdmin` is also needed when using `sky down`. Is it really necessary? If so, what is the recommended usage of SkyPilot for shared projects like ours? According to [the GCP documentation](https://cloud.google.com/iam/docs/understanding-roles?hl=fr#service-accounts-roles), the role `roles/iam.serviceAccountAdmin` bundles the following autorisations:
```
iam.serviceAccounts.create
iam.serviceAccounts.delete
iam.serviceAccounts.disable
iam.serviceAccounts.enable
iam.serviceAccounts.get
iam.serviceAccounts.getIamPolicy
iam.serviceAccounts.list
iam.serviceAccounts.setIamPolicy
iam.serviceAccounts.undelete
iam.serviceAccounts.update
resourcemanager.projects.get
resourcemanager.projects.list
```

Could you clarify which specific permissions are required for regular users to use SkyPilot?

Thank you very much. We're excited to integrate SkyPilot into our workflow!

### PR DESCRIPTION
**Pull request by Michaelvll**
This PR adds the instruction for setting up the minimal permission for GCP. Fixes #2086

This PR is done by running `sky launch` with an account having basic permissions, and gradually adding permissions according to the GCP errors.

Note: the current permission list assumes the service account already exists!

Future TODO:
- [x] Modify the `gcp/config.py` to assign less permission to the service account. #2112
- [x] Add instructions to create service account. #2112

Tested (run the relevant ones):

- [x] Code formatting: `bash format.sh`
- [x] Any manual or new tests for this PR (please specify below)
  - [x] Rendered locally
  - [x] `sky launch -c min --cloud gcp --cpus 2` with the minimal permission account.
  - [x] `sky launch -c min2 --cloud gcp --cpus 2 --num-nodes 2 -i 0` with the minimal permission account.
  - [x] `sky autostop -i 0 min`
  - [x] `sky status -r`
  - [x] `sky start min`
  - [x] `sky down min` 
  - [x] `sky launch -c min3 --cloud gcp --cpus 2 test.yaml`
  ```
  file_mounts:
    /my-mount:
      name: test-perm-filemounts
      store: gcs
      mode: mount
  run: |
    echo hi > /my-mount/user_file.txt
  ```
  - [x] `sky storage delete test-perm-filemounts`
- [x] All smoke tests: `pytest tests/test_smoke.py` with the account with minimal permission.

### CODE REVIEW

**Comment by bilelomrani1** [comment]
Thank you very much @Michaelvll @concretevitamin, I created this role in our project for further tests in our team. If you have specific commands you want me to test, don't hesitate to give me instructions.

**Comment by concretevitamin** [comment]
@bilelomrani1 Awesome! Thanks. I think a great test would be if your team could use this role to do their regular SkyPilot work. On our end we'll also run all smoke tests.

**Comment by Michaelvll** [comment]
> Sure, we can test this role for a few days and see if everything works out for us! Right off the bat, we had to add the following permissions to create a new VM:
> 
> ```
> compute.globalOperations.get
> compute.instances.setMetadata
> ```
> 
> I'll keep you updates in a few days on how things worked out for us. Many thanks for your work!

Thanks for catching this @bilelomrani1! Yes, those two permissions are required for setting up the ssh key for the user to login the instance. Just added the two permissions in the docs. : )

File: docs/source/cloud-setup/cloud-permissions.rst [hunk 1]
```diff
@@ -212,3 +212,92 @@ Optionally, to use TPUs, add the following role:
...
+    compute.zoneOperations.get
+    iam.serviceAccounts.actAs
+    iam.serviceAccounts.get
+    serviceusage.services.enable
+    serviceusage.services.list
+    serviceusage.services.use
+    resourcemanager.projects.get
+    resourcemanager.projects.getIamPolicy
+
+4. [Optional] If the user need to access GCS buckets, you can additionally add the following permissions:
```
Comment (concretevitamin) [inline docs/source/cloud-setup/cloud-permissions.rst:266]: nit: This style seems easier to read. Same below
```suggestion
4. **Optional**: If the user needs to access GCS buckets, you can additionally add the following permissions:
```

File: docs/source/cloud-setup/cloud-permissions.rst [hunk 2]
```diff
@@ -212,3 +212,92 @@ Optionally, to use TPUs, add the following role:
...
+    tpu.nodes.create
+    tpu.nodes.delete
+    tpu.nodes.list
+    tpu.nodes.get
+    tpu.nodes.update
+    tpu.operations.get
+
+5. Click **Create** to create the role.
+6. Go back to the "IAM" tab and click on **GRANT ACCESS**.
+7. Fill in the email address of the user in the “Add principles”, and select the ``minimal-skypilot-role`` in the “Assign roles” section. Click **Save**.
```
Comment (concretevitamin) [inline docs/source/cloud-setup/cloud-permissions.rst:291]:
```suggestion
7. Fill in the email address of the user in the “Add principals” section, and select ``minimal-skypilot-role`` in the “Assign roles” section. Click **Save**.
```

File: docs/source/cloud-setup/cloud-permissions.rst [hunk 3]
```diff
@@ -1,6 +1,6 @@
 .. _cloud-permissions:
 
-Minimal Cloud Permissions
+Cloud Account Permissions
```
Comment (concretevitamin) [inline docs/source/cloud-setup/cloud-permissions.rst:3]: Any reasoning behind this change?

File: sky/clouds/gcp.py [hunk 4]
```diff
@@ -575,15 +601,15 @@ def check_credentials(cls) -> Tuple[bool, Optional[str]]:
...
+
+        diffs = set(GCP_PREMISSION_CHECK_LIST).difference(set(ret_permissions))
+        if len(diffs) > 0:
             identity_str = identity[0] if identity else None
             return False, (
                 'The following permissions are not enabled for the current '
                 f'GCP identity ({identity_str}):\n    '
                 f'{diffs}\n    '
-                'For more details, visit: https://skypilot.readthedocs.io/en/latest/reference/faq.html#what-are-the-required-iam-permissons-on-gcp-for-skypilot')  # pylint: disable=line-too-long
+                'For more details, visit: https://skypilot.readthedocs.io/en/latest/cloud-setup/cloud-permissions.html#gcp')  # pylint: disable=line-too-long
```
Comment (concretevitamin) [inline sky/clouds/gcp.py:612]: Nice!

File: sky/clouds/gcp.py [hunk 5]
```diff
@@ -35,8 +35,34 @@
 # This is not a complete list but still useful to check first
```
Comment (concretevitamin) [inline sky/clouds/gcp.py:35]: This is *likely* a complete list and is useful ...

File: sky/backends/cloud_vm_ray_backend.py [hunk 6]
```diff
@@ -695,10 +695,13 @@ def _update_blocklist_on_gcp_error(
                     # This code is returned when the VM is still STOPPING.
                     self._blocked_resources.add(
                         launchable_resources.copy(zone=zone.name))
-                elif code == 8:
+                elif code in [8, 9]:
                     # Error code 8 means TPU resources is out of
                     # capacity. Example:
                     # {'code': 8, 'message': 'There is no more capacity in the zone "europe-west4-a"; you can try in another zone where Cloud TPU Nodes are offered (see https://cloud.google.com/tpu/docs/regions) [EID: 0x1bc8f9d790be9142]'} # pylint: disable=line-too-long
+                    # Error code 9 means TPU resources is insufficient reserved
```
Comment (concretevitamin) [inline sky/backends/cloud_vm_ray_backend.py:702]: Nice! cc @infwinston

**Review by concretevitamin — APPROVED** [review]
LGTM, thanks @Michaelvll! This would help onboarding new organizations a lot.

File: sky/clouds/gcp.py (hunk 5 above)
Comment (Michaelvll) [reply sky/clouds/gcp.py:35]: Good catch! Fixed. Thanks!

File: docs/source/cloud-setup/cloud-permissions.rst (hunk 3 above)
Comment (Michaelvll) [reply docs/source/cloud-setup/cloud-permissions.rst]: Ahh, this was a remnant. Reverted. Thanks!

**Pull request merged**